From 1f0746f038a6f65cc4c06581bc1c5942188bd0a8 Mon Sep 17 00:00:00 2001
From: haiqiang yao <haiqiangx.yao@intel.com>
Date: Sun, 4 Jan 2015 16:23:59 +0800
Subject: [PATCH 202/210] atomisp2: Add zoom region reasonable justification

the patch will add zoom region reasonable justification when set
per frame setting to css. it will fix regression 1403972158 which
is "CTS Verifier Camera Formats/Video  preview random stretch or
corruption when switch camera resolution"

Change-Id: I89781428502fffb7153c3935135222f6ac356146
Tracked-On:
https://hsdes.intel.com/home/default.html#article?id=1403972158
Signed-off-by: haiqiang yao <haiqiangx.yao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/34849
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_cmd.c      | 24 +++++++++++++++++-----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index 817c472..d0b103d 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -2760,6 +2760,14 @@ static bool atomisp_check_zoom_region(
 
 	if ((w <= config.width) && (h <= config.height) && w > 0 && h > 0)
 		flag = true;
+	else
+		/* setting error zoom region */
+		dev_err(asd->isp->dev, "%s zoom region ERROR:dz_config:origin (%d,%d),region(%d,%d)eff_res(%d, %d)\n",
+				__func__, dz_config.zoom_region.origin.x,
+				dz_config.zoom_region.origin.y,
+				dz_config.zoom_region.resolution.width,
+				dz_config.zoom_region.resolution.height,
+				config.width, config.height);
 
 	return flag;
 }
@@ -2895,11 +2903,17 @@ static int __atomisp_cp_general_isp_parameters(
 				   sizeof(struct atomisp_css_dp_config)))
 			return -EFAULT;
 	if (asd->run_mode->val != ATOMISP_RUN_MODE_VIDEO) {
-		if (arg->dz_config &&
-			atomisp_check_zoom_region(asd, arg) == true)
-			if (copy_from_user(&css_param->dz_config, arg->dz_config,
-				   sizeof(struct atomisp_css_dz_config)))
-				return -EFAULT;
+		if (arg->dz_config) {
+			if (atomisp_check_zoom_region(asd, arg)) {
+				if (copy_from_user(&css_param->dz_config,
+					arg->dz_config,
+					sizeof(struct atomisp_css_dz_config)))
+					return -EFAULT;
+			} else {
+				dev_err(asd->isp->dev, "crop region error!");
+				return -EINVAL;
+			}
+		}
 	}
 
 	if (arg->nr_config)
-- 
1.9.1

